content/updates/steps-to-break-up-a-terralith.md

```diff
@@ -0,0 +1,361 @@
+
+- [Overview](#overview)
+- [Why Is It a Problem?](#why-is-it-a-problem)
+- [Before You Get Started — What You Need to Think About](#before-you-get-started--what-you-need-to-think-about)
+  - [The Importance of Version Selection in Terralith Migration](#the-importance-of-version-selection-in-terralith-migration)
+  - [Identify Logical Groupings](#identify-logical-groupings)
+  - [Assess Dependencies](#assess-dependencies)
+  - [State Management](#state-management)
+  - [Key Organizational and Team Questions](#key-organizational-and-team-questions)
+- [Migrating to Modular Terraform](#migrating-to-modular-terraform)
+  - [#1 Confirm a Clean Plan](#1-confirm-a-clean-plan)
+  - [#2 Create a Backup](#2-create-a-backup)
+  - [#3 Create New Module Structures](#3-create-new-module-structures)
+  - [#4 Inter-Module Communication (if any)](#4-inter-module-communication-if-any)
+  - [#5 Move Resources to The New State](#5-move-resources-to-the-new-state)
+  - [#6 Validate the New Modules](#6-validate-the-new-modules)
+- [The Final Checks and Tips](#the-final-checks-and-tips)
+- [Grab Our Terralith Breakup Checklist](#grab-our-terralith-breakup-checklist)
+- [Conclusion](#conclusion)
+
+## Overview
+
+Do you manage your infrastructure with Terraform or OpenTofu (collectively referred to from here on out as TF)? Have you noticed long `plan` and `apply` durations? Or are you concerned about the blast radius of changes? If so, you might be dealing with a **Terralith**. “Terralith” refers to a monolithic TF architecture where all your resources are managed within a single root module and state file. This setup leads to inefficiencies and increased risk when making changes.
+
+In our previous article, [“What Is a Terralith?”](https://masterpoint.io/updates/terralith-monolithic-terraform-architecture/), we explored the challenges associated with Terraliths and why it’s beneficial to break them into smaller, manageable pieces.
+
+In this guide, we’ll walk you through the steps to break up a Terralith, helping you improve your infrastructure’s maintainability, and overall performance.
+
+## Why Is It a Problem?
+
+The Terralith anti-pattern is insidious. It might not cause immediate disruptions, but that single root module grows behind the scenes. Eventually, it snowballs into a significant challenge that hits hard when you least expect it. The longer you wait to address it, the more complex and time-consuming the solution becomes.
+
+Let’s take a look at the typical symptoms of a Terralith:
+- **Large state file(s)**: A large number of resources are stuffed into a single `.tfstate` state file. There’s no universally correct file size or resource count for TF. However, our experience suggests that **between 30 and 100 resources per state file** is the range that strikes a good balance between optimal performance and manageability. Ideally, your TF root modules around service boundaries that share the same lifecycle, allowing for easy division.
```

Comment:
I'm genuinely worried about the pushback on this line. People are going to freak out and hate this I think. Mind if we change this?

```suggestion
- **Large state file(s)**: A large number of resources are stuffed into a single `.tfstate` state file. There’s no universally correct file size or resource count for TF. However, our experience suggests that **if your state files have between a few hundred and 1000+ resources** then you're likely dealing with performance and manageability issues. Not all APIs are the same, so that might not be the case for you, but we're painting with a broad brush here to give you an idea. Ideally, your TF root modules are built around service boundaries that share the same lifecycle, allowing for easy division.
```

Reply:
Dan's suggestion was to give some specific numbers, that's why we have this line. Considering recent talks about Terralliths... I agree with you here.
I think it's a great addition about the APIs btw 👍